DockDoor/Utilities/WindowUtil.swift

```diff
@@ -492,85 +529,148 @@ final class WindowUtil {
-            for task in tasks {
-                group.addTask {
-                    try await task.value
+        if result == .success, let windows = value as? [AXUIElement] {
+            desktopSpaceWindowCacheManager.updateCache(bundleId: bundleID) { cachedWindows in
+                for window in windows {
+                    let isMinimized: Bool = getAXAttribute(element: window, attribute: kAXMinimizedAttribute as CFString) ?? false
+                    let windowName: String? = getAXAttribute(element: window, attribute: kAXTitleAttribute as CFString)
+                    
+                    if isMinimized || isParentAppHidden, let windowName = windowName {
+                        cachedWindows = Set(cachedWindows.map { windowInfo in
+                            var updatedWindow = windowInfo
+                            if windowInfo.windowName == windowName {
+                                updatedWindow.isMinimized = isMinimized
+                                updatedWindow.isHidden = isParentAppHidden
+                            }
+                            return updatedWindow
+                        })
+                    }
+                }
+                
+                if isParentAppHidden {
+                    cachedWindows = Set(cachedWindows.map { windowInfo in
+                        var updatedWindow = windowInfo
+                        updatedWindow.isHidden = true
+                        return updatedWindow
+                    })
                 }
             }
-            
-            var results: [T] = []
-            for try await result in group {
-                results.append(result)
-            }
-            return results
         }
     }
-}
-
-actor AsyncSemaphore {
-    private var value: Int
-    private var waiters: [CheckedContinuation<Void, Never>] = []
     
-    init(value: Int) {
-        self.value = value
+    private static func getNonLocalizedAppName(forBundleIdentifier bundleIdentifier: String) -> String? {
+        guard let bundleURL = NSWorkspace.shared.urlForApplication(withBundleIdentifier: bundleIdentifier) else {
+            return nil
+        }
+        
+        let bundle = Bundle(url: bundleURL)
+        return bundle?.object(forInfoDictionaryKey: "CFBundleName") as? String
+    }
+    
+    private static func updateAppNameBundleIdTracker(app: SCRunningApplication, nonLocalName: String) {
+        appNameBundleIdTracker[app.applicationName] = app.bundleIdentifier
+        appNameBundleIdTracker[nonLocalName] = app.bundleIdentifier
     }
     
-    func wait() async {
-        if value > 0 {
-            value -= 1
-        } else {
-            await withCheckedContinuation { continuation in
-                waiters.append(continuation)
-            }
+    static func findAllWindowsInDesktopCacheForApplication(for applicationName: String) -> [WindowInfo]? {
+        let bundleID = findBundleID(for: applicationName)
```

Comment:
hmm, I was not aware that applications that spawn windows are allowed to exist without bundle ids - I thought all applications needed a bundle id. Do you happen to know of applications that exist in our day-to-day lives that I can use to append/alter this logic that don't happen to have a bundle id?

Reply:
Yes it's pretty rare, I know about android studio emulator and scrcpy

Reply:
Now that you mention this, I recall reading some comments in Alt-Tab about this. Was it just the emulator or the whole application? I am going to download it and do some testing with it to see the behavior and what else we might need to enhance.

Reply:
only the emulator

Reply:
Interesting ok. Let me download Android Studio and let me compare how the native OS is reacting. Depending on what the OS shows, ill try my best to match the same outcome

Reply:
Note that at https://github.com/ejbills/DockDoor/pull/214 I changed the approach to be more reliable in general (we work with `NSRunningApplication`)

Reply:
Apologies this took a little while to get done, attached is a screenshot where we are able to pick up on the emulator and the IDE window as separate windows. I suggest we move forward with the bundle identifier, I am unaware of applications that can be compiled without a bundle Identifier for the apple ecosystem.
<img width="1508" alt="Screenshot 2024-08-01 at 13 57 50" src="https://github.com/user-attachments/assets/406c330c-ead3-44b7-8b8d-c9842f634028">

Reply:
This is on Mac OS Dev Beta 15.1 with Android Studio Koala. Not my Android project or anything, just one of the default ones they had

Reply:
This is not an Android Studio emulator window. This is an Android Studio window. The Android Studio emulator is a completely separate floating window, you dragged out a part of Android Studio, "Running Devices". Also check out scrcpy.

Other than that, I don't see why we need to base on bundleID and convert back and forth to the appropriate information, when we have a much juicier information object like NSRunningApplication.

Reply:
Maybe put my change in the current under mouse app detection method into this PR? This way we can work with NSRunningApplication all the way through.

Reply:
https://developer.android.com/studio/run/emulator-launch-separate-window#:~:text=To%20launch%20the%20emulator%20in,the%20Running%20Devices%20tool%20window.

As of 2024-07-01 that is the Android Studio Emulator window. I also recall in the past that it would spawn a different style of window but I believe they have changed things around.

Reply:
To get that information object ```NSRunningApplication``` we need to make another call to get that information. We have the right information we need to perform the store/save operation and hence we don't need to make potentially another system call for information we already have. Storing the information is likely cheaper than making a system call because the OS is most likely keeping the application in the Cache. 

@ejbills correct me if I am wrong too.

Reply:
@ShlomoCode I don't have an Android device to fidget with scrcpy. Any chance you can pull the latest changes and see if it's not recognizing the window?

Reply:
@ShlomoCode this is the emulator window. I tried to just launch it on an older version of Android Studio, I believe a June 2023 Giraffe release and still the same thing.

<img width="1512" alt="Screenshot 2024-08-02 at 10 37 12" src="https://github.com/user-attachments/assets/351d6902-7b03-4d9e-a04a-4cb0db42d0e0">

Reply:
> @ShlomoCode I don't have an Android device to fidget with scrcpy.

scrcpy can connect to android studio emulator

Reply:
I pulled the PR and checked and it does not recognize a scrcpy window, even when it is not set to float above everything.

Reply:
>However, some emulator features are only available when you run it in a separate window. To launch the emulator in a separate window, go to...

***

> To get that information object `NSRunningApplication` we need to make another call to get that information

Why?
```
guard let runningApp = NSRunningApplication.runningApplications(withBundleIdentifier: bundleIdentifier).first else {
    print("Current App is not running (\(bundleIdentifier))")
    return nil
}
```

Reply:
Since the current prod version of the app doesn't support apps without a bundle ID, I think this can be made as an enhancement in the future. For now, I will be merging the PR to get some users this feature and get testing going.

Reply:
Makes sense